egui-wgpu/src/winit.rs

```diff
@@ -37,36 +75,131 @@ impl Painter {
         ))
         .unwrap();
 
-        let size = window.inner_size();
-        let surface_format = surface.get_preferred_format(&adapter).unwrap();
-        let surface_config = wgpu::SurfaceConfiguration {
-            usage: wgpu::TextureUsages::RENDER_ATTACHMENT,
-            format: surface_format,
-            width: size.width as u32,
-            height: size.height as u32,
-            present_mode: wgpu::PresentMode::Fifo, // TODO: make vsync configurable
-        };
-        surface.configure(&device, &surface_config);
-
-        let egui_rpass = renderer::RenderPass::new(&device, surface_format, msaa_samples);
+        let egui_rpass = renderer::RenderPass::new(&device, target_format, self.msaa_samples);
 
-        Self {
+        RenderState {
             device,
             queue,
-            surface_config,
-            surface,
+            target_format,
             egui_rpass,
         }
     }
 
-    pub fn max_texture_side(&self) -> usize {
-        self.device.limits().max_texture_dimension_2d as usize
+    // We want to defer the initialization of our render state until we have a surface
+    // so we can take its format into account.
+    //
+    // After we've initialized our render state once though we expect all future surfaces
+    // will have the same format and so this render state will remain valid.
+    fn ensure_render_state_for_surface(&mut self, surface: &Surface) {
+        if self.adapter.is_none() {
+            let adapter =
+                pollster::block_on(self.instance.request_adapter(&wgpu::RequestAdapterOptions {
+                    power_preference: self.power_preference,
+                    compatible_surface: Some(surface),
+                    force_fallback_adapter: false,
+                }))
+                .unwrap();
+
+            self.adapter = Some(adapter);
+        }
```

Comment:
I just tried this out but it's doesn't seem to work out well in terms of borrow checking...

`get_or_insert_with` returns a mutable reference which we don't really want here and that ends up indirectly holding a mutable reference to `self` that then means the adapter initialization needs to be wrapped in a redundant block to make sure the borrow checker will let us ensure the render_state afterwards. Even trying to downgrade the adapter reference to immutable afterwards doesn't work since it's still based on the mutable reference to self.

I guess an API conceptually like `.as_ref_or_insert_with(|| ..)` for setting up the adapter would have been convenient (which wouldn't take a mutable reference), and then the adapter reference could have been used to set up the render_state too.